### PR TITLE
Exactly Once Delivery and Transactional Messaging Support

### PR DESCRIPTION
**Pull request by nguyenquangminh0711**
One of the most important features of Kafka 0.11 is the support of "Exactly Once Delivery" (Idempotent message production) and Transactional Messaging. This PR is to add the supports for those features, based on [the official documentation](https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit#heading=h.xq0ee1vnpz4o).

The Exactly Once Delivery adds a semantic to Kafka that it prevents the message duplication in the situations that the producer re-send the same messages after a network problem, broker outage, etc. Kafka maintains two pair of parameters: ProducerID and Sequence ID. When the client produces messages, those two parameters are attached in the produce request. If the request succeeds, both client and broker updates the current sequence number of the producer ID for each partition. If the client sends some messages having smaller sequence than the broker's record, Kafka assumes those messages are duplication and simply ignore.

The implementation of this feature is quite simple. First, before any produce requests are triggered, the client triggers a request to any brokers to fetch a *new* Producer ID, and use it in later produce requests. To use the feature, one option needs to be added in the producer declaration. And the async producer is supported out of the box.

```ruby
producer = kafka.producer(idempotent: true)
producer.produce('Testing', topic: 'test')
producer.deliver_messages

producer = kafka.async_producer(idempotent: true)
producer.produce('Testing 2', topic: 'test')
```

This option is off by default. However, I'm thinking of switch it on by default, since it adds up just a little bit overhead, but the benefits gains are huge. The normal users won't mind about turning it off. What do you think @dasch?

The pitfalls when using idempotent producing is that it is only effective on a single producer session. If the producer is off and on again, it is granted new Producer ID and all sequential messages are considered new. To ensure idempotent over multiple sessions, we could pass `transactional_id` so that the old producer with the same transactional_id is fenced out. This feature is advanced and only used with care.

```ruby
producer = kafka.producer(idempotent: true, transactional_id: 'some_machine_id')
producer.produce('Testing', topic: 'test')
producer.deliver_messages
```

One more thing, exactly once delivery is from the Kafka side only. If the client fails to handle such exceptions, it is likely that it would call producing twice and the duplication still exists.

The transactional guarantees are to ensure that messages delivered to partitions belonging to different brokers are treated as a single unit; if one fails, other fails and vice vera. It fixes the current situation that each produce operation could send many requests to leader brokers of the partitions. If one of them fails after a successful one, those messages can not be undone. In addition, it allows us to make messages to different topics atomic. That's a really game changer.

To achieve this, Kafka requires a TransactionalID in each produce request. The messages are recorded but invisible to the consumers, unless the consumers turns on dirty read option. At the end of the transaction, the client sends EndTxn requests (whether Commit / Abort) to close the transaction. The request is hang until all brokers agree about the result and clean up the transaction logs. It is highly recommended to read the above documentation.

The proposed usage of the transactional producer:

```ruby
producer = kafka.producer(transactional: true, transactional_id: 'someid')
producer.init_transactions
producer.begin_transaction
producer.produce('a', topic: 'test')
producer.deliver_messages
producer.produce('b', topic: 'test2')
producer.deliver_messages
producer.commit_transaction # Or producer.abort_transaction
```

Or with some syntax sugar:

```ruby
producer = kafka.producer(transactional: true, transactional_id: 'someid')
producer.init_transactions
producer.transaction do
  producer.produce('a', topic: 'test')
  producer.produce('b', topic: 'test2')
  producer.deliver_messages
end

producer.transaction do
  producer.produce('a', topic: 'test')
  raise Kafka::Producer::AbortTransaction
end
```

I haven't finished the implementation of this feature yet. I'll hang here for discussion and progress update.

## Things to do
- [x] Add InitProducerId API
- [x] Implement idempotent producer
- [x] Functional tests for idempotent producer & async producer
- [x] Implement TransactionManager and state machine to manage transaction
- [x] Integrate and implement transaction flow
- [x] Tests for happy scenarios
- [x] Tests for transaction error scenarios
- [x] Tests for idempotent combined with transaction production
- [x] Handle abort transactions in client side.
- [x] Tests for edge cases (timeout, producer fenced out, broker down, etc.)
- [x] ~Make client more fault tolerant, handle and auto-recover from common problems (timeout retrying, message too large, etc.)~
- [x] ~Support message processing before transaction is finished. Dirty read, AddOffsetsToTxn, TxnOffsetCommit, etc.~
- [x] Write documents (tons of them), and examples

### CODE REVIEW

**Comment by nguyenquangminh0711** [comment]
@dasch I think you can start the review. I covered almost all the cases of idempotent and transactional producer. There are some tests to add and some documents to change, but the main implementation is done. I think I would postpone the dirty read flow that allows the consumers to consume uncommitted transactions to later phase.

**Comment by nguyenquangminh0711** [comment]
@dasch I fulfill the tests, as well as the documents for the implementation. It looks it is ready for a beta release. Please have a look 💪 I'm just not sure about the interfaces of the transaction. Do we need an easier and more friendly interface?

Also, I have a benchmark on my machine. The support for idempotent / transactional producer doesn't slow down the overall processing time, although the 0.7 beta releases are slow, comparing to the current stable branch. I'm investigating and would provide an enhancement soon.

```
0.6.7 - 1000000 messages - 100 bytes each => 54.71287566667888
0.7.0.beta1 - 1000000 messages - 100 bytes each => 93.92053633334581
---
This branch:
- Normal producer - 1000000 messages - 100 bytes each => 85.6134163333724
- Idempotent producer  - 1000000 messages - 100 bytes each => 88.89357166659708
- Transactional producer - 1000000 messages - 100 bytes each => 86.84967833330545
```

File: lib/kafka/transaction_manager.rb [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+      @logger = logger
+
+      @transactional = transactional
+      @transactional_id = transactional_id
+      @transactional_timeout = transactional_timeout
+      @transaction_state = Kafka::TransactionStateMachine.new(logger: logger)
+      @transaction_partitions = {}
+
+      # If transactional mode is enabled, idempotent must be enabled
+      @idempotent = transactional ? true : idempotent
```
Comment (dasch) [inline lib/kafka/transaction_manager.rb]:
```ruby
@idempotent = transactional || idempotent
```

File: lib/kafka/transaction_state_machine.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    def committing_transaction?
+      @mutex.synchronize { @state == COMMITTING_TRANSACTION }
+    end
+
+    def aborting_transaction?
+      @mutex.synchronize { @state == ABORTING_TRANSACTION }
+    end
+
+    def error?
+      @mutex.synchronize { @state == ERROR }
```
Comment (dasch) [inline lib/kafka/transaction_state_machine.rb]: Just to avoid future errors, can we extract a private helper method here? e.g.

```ruby
def error?
  in_state?(ERROR)
end

private

def in_state?(state)
  @mutex.synchronize { @state == state }
end
```

File: lib/kafka/cluster.rb [hunk 3]
```diff
@@ -310,8 +293,7 @@ def resolve_offsets(topic, partitions, offset)
             topic => broker_partitions.map {|partition|
               {
                 partition: partition,
-                time: offset,
-                max_offsets: 1,
+                time: offset
```
[the remark this answers is not in the document]
Comment (nguyenquangminh0711) [reply lib/kafka/cluster.rb]: @dasch I bumped the version of ListOffset API so that it supports the Isolation Level. The new version of this API drops the support of `max_num_offsets`. It returns the latest offset, instead of an array.

[the remark this answers is not in the document]
Comment (dasch) [reply lib/kafka/cluster.rb]: :+1:

**Comment by dasch** [comment]
Would it be possible to run this branch in a production setting for a week or so before I merge and release an alpha?

**Comment by nguyenquangminh0711** [comment]
@dasch I will deploy to my company's staging environment. It has nearly the same configuration with the production, but less traffic of course. I'll apply into our applications, as well as setup some scenarios to test. I am not allowed to test on production environment though 🤣

**Comment by nguyenquangminh0711** [comment]
@dasch I deployed this branch to my staging environment and have a lot of tests, with verification in both producers and consumers to ensure all the functions work as expected, total 50M messages tested. During the tests, I found some bugs and already resolved. I think that's stable enough to go now 💪

**Comment by dasch** [comment]
Awesome! :heart:

I'll release v0.7.0 first in order to have a stable base, but I will merge this PR this week :+1:

**Comment by nguyenquangminh0711** [comment]
Hi @dasch, any updates on this? Is it ready to merge? Or are you waiting for something?

**Comment by dasch** [comment]
:heart:

I was letting v0.7.0 run in our production environment for a little while to ascertain that no obvious problems were introduced before merging this, as having multiple big changes deployed in one go makes it harder to track down bugs.

I'll release v0.7.1.beta1 now, please test it out in production!

**Comment by mensfeld** [comment]
@nguyenquangminh0711 amazing piece of work!

**Comment by Elyahou** [comment]
@nguyenquangminh0711 Is it possible to commit offset of the consumer in the same transaction as of a message produce call ?
